AX: Collapse AccessibilityTable into AccessibilityRenderObject / AccessibilityNodeObject

twilco opened the pull request:
#### 73a1e672ef8e96a43f43906f3d5d372ebf128e09
<pre>
AX: Collapse AccessibilityTable into AccessibilityRenderObject / AccessibilityNodeObject
<a href="https://rdar.apple.com/158610397">rdar://158610397</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=297540">https://bugs.webkit.org/show_bug.cgi?id=297540</a>

Reviewed by Joshua Hoffman.

Whether or not something is a table can change depending on the role or renderer it has. Today we deal with this by
destroying and recreating the object, which harms performance by forcing us to create new AXIsolatedObjects, update
children of other objects, etc. It&apos;s also harmful to ATs, who lose what was otherwise a valid reference to a UIElement,
which can cause various bad behavior such as unexpected focus / AT cursor movement. Finally, we only do this when the
role changes, not when the renderer changes, so the workaround is not comprehensive.

With this commit, we no longer have this concern, as AccessibilityTable is deleted, and functionality moved into
AccessibilityRenderObject / AccessibilityNodeObject.

This commit adds AXObjectRareData to hold onto table-related state. This follows pretty closely to the pattern of
NodeRareData.

* LayoutTests/accessibility/mac/aria-grid-with-strange-hierarchy-expected.txt:
* LayoutTests/accessibility/mac/aria-grid-with-strange-hierarchy.html:
Change expected behavior to include the first row, which is a better experience for ATs and a valid interpretation of
this markup.
* LayoutTests/accessibility/multiple-attribute-change-crash-expected.txt:
* LayoutTests/accessibility/multiple-attribute-change-crash.html:
* Source/WebCore/SaferCPPExpectations/UncheckedCallArgsCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncheckedLocalVarsCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncountedLambdaCapturesCheckerExpectations:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/accessibility/AXCoreObject.cpp:
(WebCore::AXCoreObject::unignoredChildren):
(WebCore::AXCoreObject::firstUnignoredChild):
(WebCore::AXCoreObject::selectedRows):
(WebCore::AXCoreObject::columnHeader):
* Source/WebCore/accessibility/AXCoreObject.h:
(WebCore::Accessibility::exposedTableAncestor):
* Source/WebCore/accessibility/AXLogger.cpp:
(WebCore::AXLogger::log):
(WebCore::operator&lt;&lt;):
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::createObjectFromRenderer):
(WebCore::AXObjectCache::createFromNode):
(WebCore::AXObjectCache::handleChildrenChanged):
(WebCore::AXObjectCache::handleRecomputeCellSlots):
(WebCore::AXObjectCache::handleRowCountChanged):
(WebCore::AXObjectCache::handleRoleChanged):
(WebCore::AXObjectCache::handleAttributeChange):
(WebCore::AXObjectCache::performDeferredCacheUpdate):
(WebCore::AXObjectCache::deferRecomputeTableCellSlots):
(WebCore::isAccessibilityTable): Deleted.
(WebCore::isAccessibilityARIATable): Deleted.
(WebCore::isTableOrRowRole): Deleted.
* Source/WebCore/accessibility/AXObjectCache.h:
* Source/WebCore/accessibility/AXObjectRareData.cpp: Added.
(WebCore::AXObjectRareData::resetChildrenDependentTableFields):
* Source/WebCore/accessibility/AXObjectRareData.h: Added.
(WebCore::AXObjectRareData::tableRows const):
(WebCore::AXObjectRareData::tableColumns const):
(WebCore::AXObjectRareData::rowCount const):
(WebCore::AXObjectRareData::columnCount const):
(WebCore::AXObjectRareData::appendColumn):
(WebCore::AXObjectRareData::appendRow):
(WebCore::AXObjectRareData::isExposableTable const):
(WebCore::AXObjectRareData::setIsExposableTable):
(WebCore::AXObjectRareData::tableHeaderContainer const):
(WebCore::AXObjectRareData::setTableHeaderContainer):
(WebCore::AXObjectRareData::cellSlots const):
(WebCore::AXObjectRareData::mutableCellSlots):
* Source/WebCore/accessibility/AXSearchManager.cpp:
(WebCore::AXSearchManager::matchForSearchKeyAtIndex):
(WebCore::appendChildrenToArray):
* Source/WebCore/accessibility/AXTableHelpers.cpp: Added.
(WebCore::AXTableHelpers::appendCaptionTextIfNecessary):
(WebCore::AXTableHelpers::isTableRole):
(WebCore::AXTableHelpers::tableElementIncludingAncestors):
(WebCore::AXTableHelpers::tableElementIndicatesAccessibleTable):
(WebCore::AXTableHelpers::tableSectionIndicatesAccessibleTable):
(WebCore::AXTableHelpers::styleFrom):
(WebCore::AXTableHelpers::isDataTableWithTraversal):
* Source/WebCore/accessibility/AXTableHelpers.h: Added.
* Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
(WebCore::AccessibilityNodeObject::init):
(WebCore::AccessibilityNodeObject::determineAccessibilityRole):
(WebCore::AccessibilityNodeObject::determineAccessibilityRoleFromNode const):
(WebCore::AccessibilityNodeObject::clearChildren):
(WebCore::AccessibilityNodeObject::addChildren):
(WebCore::AccessibilityNodeObject::isMultiSelectable const):
(WebCore::AccessibilityNodeObject::cellForColumnAndRow):
(WebCore::AccessibilityNodeObject::rareDataWithCleanTableChildren):
(WebCore::AccessibilityNodeObject::cells):
(WebCore::AccessibilityNodeObject::columnCount):
(WebCore::AccessibilityNodeObject::rowCount):
(WebCore::AccessibilityNodeObject::cellSlots):
(WebCore::AccessibilityNodeObject::axRowCount const):
(WebCore::AccessibilityNodeObject::axColumnCount const):
(WebCore::AccessibilityNodeObject::updateRowDescendantRoles):
(WebCore::AccessibilityNodeObject::setCellSlotsDirty):
(WebCore::AccessibilityNodeObject::headerContainer):
(WebCore::AccessibilityNodeObject::columns):
(WebCore::AccessibilityNodeObject::rows):
(WebCore::AccessibilityNodeObject::isDataTable const):
(WebCore::AccessibilityNodeObject::rowHeaders):
(WebCore::AccessibilityNodeObject::visibleRows):
(WebCore::AccessibilityNodeObject::addTableChildrenAndCellSlots):
(WebCore::AccessibilityNodeObject::computeCellSlots):
(WebCore::AccessibilityNodeObject::recomputeIsExposableIfNecessary):
(WebCore::AccessibilityNodeObject::isTable const):
(WebCore::AccessibilityNodeObject::labelText const):
(WebCore::shouldUseAccessibilityObjectInnerText):
* Source/WebCore/accessibility/AccessibilityNodeObject.h:
(WebCore::AccessibilityNodeObject::computeIsTableExposableThroughAccessibility const):
* Source/WebCore/accessibility/AccessibilityObject.cpp:
(WebCore::AccessibilityObject::init):
(WebCore::AccessibilityObject::insertChild):
(WebCore::AccessibilityObject::ensureRareData):
* Source/WebCore/accessibility/AccessibilityObject.h:
(WebCore::AccessibilityObject::hasRareData const):
(WebCore::AccessibilityObject::rareData const):
(WebCore::AccessibilityObject::needsRareData const):
(WebCore::AccessibilityObject::isAccessibilitySVGObjectInstance const):
(WebCore::AccessibilityObject::isAriaTable const):
(WebCore::AccessibilityObject::recomputeIsExposableIfNecessary):
(WebCore::AccessibilityObject::ensureCellIndexesUpToDate):
* Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
(WebCore::AccessibilityRenderObject::computeIsIgnored const):
(WebCore::AccessibilityRenderObject::determineAccessibilityRole):
(WebCore::AccessibilityRenderObject::addChildren):
* Source/WebCore/accessibility/AccessibilityTable.cpp: Removed.
* Source/WebCore/accessibility/AccessibilityTable.h: Removed.
* Source/WebCore/accessibility/AccessibilityTableCell.cpp:
(WebCore::AccessibilityTableCell::parentTable const):
(WebCore::AccessibilityTableCell::isExposedTableCell const):
(WebCore::AccessibilityTableCell::determineAccessibilityRole):
* Source/WebCore/accessibility/AccessibilityTableCell.h:
* Source/WebCore/accessibility/AccessibilityTableColumn.cpp:
(WebCore::AccessibilityTableColumn::addChildren):
* Source/WebCore/accessibility/AccessibilityTableHeaderContainer.cpp:
(WebCore::AccessibilityTableHeaderContainer::addChildren):
* Source/WebCore/accessibility/AccessibilityTableHeaderContainer.h:
* Source/WebCore/accessibility/AccessibilityTableRow.cpp:
(WebCore::AccessibilityTableRow::isTableRow const):
(WebCore::AccessibilityTableRow::parentTable const):
(WebCore::AccessibilityTableRow::disclosedRows):
(WebCore::AccessibilityTableRow::disclosedByRow const):
* Source/WebCore/accessibility/AccessibilityTableRow.h:
* Source/WebCore/accessibility/ios/WebAccessibilityObjectWrapperIOS.mm:
(-[WebAccessibilityObjectWrapper tableParent]):
(-[WebAccessibilityObjectWrapper accessibilityHeaderElements]):
(-[WebAccessibilityObjectWrapper accessibilityElementForRow:andColumn:]):
(-[WebAccessibilityObjectWrapper accessibilityRowCount]):
(-[WebAccessibilityObjectWrapper accessibilityColumnCount]):
(-[WebAccessibilityObjectWrapper accessibilityARIARowCount]):
(-[WebAccessibilityObjectWrapper accessibilityARIAColumnCount]):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.cpp:
(WebCore::AXIsolatedObject::isAccessibilityTableInstance const): Deleted.
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.h:
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:
(WebCore::AXIsolatedTree::updateDependentProperties):
(WebCore::createIsolatedObjectData):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.h:
* Source/WebCore/accessibility/mac/AXObjectCacheMac.mm:
(WebCore::AXObjectCache::postPlatformNotification):
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperBase.mm:
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.mm:
(-[WebAccessibilityObjectWrapper _additionalAccessibilityAttributeNames:]):
(-[WebAccessibilityObjectWrapper accessibilityAttributeNames]):
(-[WebAccessibilityObjectWrapper accessibilityAttributeValue:]):
(-[WebAccessibilityObjectWrapper accessibilityParameterizedAttributeNames]):
(-[WebAccessibilityObjectWrapper _accessibilitySetValue:forAttribute:]):
(-[WebAccessibilityObjectWrapper accessibilityAttributeValue:forParameter:]):

Canonical link: <a href="https://commits.webkit.org/298972@main">https://commits.webkit.org/298972@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b67ce8ac4347883f6ff09bd579f6b9456745e97a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/117312 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/36982 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/27598 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/123411 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/69298 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/4bdb6203-a1cb-4f47-bdbd-15276b518e1b) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/119187 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/37678 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/45569 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/89029 "Passed tests") | [💥 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/43701 "An unexpected error occured. Recent messages:Printed configuration") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/a814cf41-21d5-4985-913b-dd95463129e9) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/120245 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/29995 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/105188 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/69536 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/31c8dcd7-9ca6-4242-abee-8e339db246cf) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/29052 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/23301 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/67084 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/99396 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/23485 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/126532 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/44209 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/33219 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/97699 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/44565 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/101423 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/97494 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24827 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/42849 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/20779 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/40559 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/44082 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/49741 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/43538 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/46883 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/45234 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

hoffmanjoshua commented on Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
```diff
@@ -132,6 +138,10 @@ void AccessibilityNodeObject::init()
     ASSERT(!m_initialized);
     m_initialized = true;
 #endif
+    m_ariaRole = determineAriaRoleAttribute();
```
Will this cause us to do the `determineAriaRoleAttribute()` work twice? In most (all?) `determineRoleValue` overrides we set `m_ariaRole`.

twilco replied:
Great point! I fixed this in the latest commit so we do it once in `init()`, then only do it again after the `role` attribute actually changes.

hoffmanjoshua commented on Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
```diff
@@ -1738,6 +1780,608 @@ bool AccessibilityNodeObject::isGenericFocusableElement() const
+
+    if (!isExposableTable()) [[unlikely]]
+        return;
+    unsigned desiredColumnCount = computeCellSlots();
```
I know this was already there before, but would be nice if we computed the cell slots after null-checking the cache below. Especially since `computeCellSlots` already does a null check for `axObjectCache`

twilco replied:
Fixed!

hoffmanjoshua commented on Source/WebCore/accessibility/AccessibilityNodeObject.h:
```diff
@@ -78,6 +81,36 @@ class AccessibilityNodeObject : public AccessibilityObject {
+    AccessibilityChildrenVector visibleRows() final;
+
+    // Returns an object that contains, as children, all the objects that act as headers.
+    AccessibilityObject* headerContainer() final;
```
Do you think we should rename this to `tableHeaderContainer`, since it just calls the rare data method of the same name? Especially with us removing the `AccessibilityTable` class.

hoffmanjoshua commented on Source/WebCore/accessibility/AccessibilityTableCell.cpp:
```diff
@@ -85,16 +84,15 @@ bool AccessibilityTableCell::computeIsIgnored() const
-            return ancestorTable && ancestorTable->isExposable() && !ancestorTable->isIgnored();
-        }));
+        return Accessibility::findAncestor<AccessibilityObject>(*this, false, [] (const auto& ancestor) {
+            return ancestor.isExposableTable() && !ancestor.isIgnored();
```
Checking `isIgnored` shouldn't be necessary here right? Or should we assert somewhere that if the table is exposable, that it isn't ignored?

twilco replied:
I think when considering things like `aria-hidden`, it _is_ possible and valid to have an ignored, exposable table 🤔

hoffmanjoshua commented on Source/WebCore/accessibility/AXTableHelpers.cpp:
```diff
@@ -0,0 +1,341 @@
+    return element.existingComputedStyle();
+}
+
+bool isDataTableWithTraversal(HTMLTableElement& tableElement, const RenderStyle* tableStyle, AXObjectCache& cache)
```
Is the `tableStyle` something we can get from the `HTMLTableElement`, so we don't need to pass them in separately?

twilco replied:
Fixed!

hoffmanjoshua reviewed (approved):
I think this looks really good! Just have a few questions to see if we can simplify/clarify some things now that the classes are merged

twilco committed:
AX: Collapse AccessibilityTable into AccessibilityRenderObject / AccessibilityNodeObject
rdar://158610397
https://bugs.webkit.org/show_bug.cgi?id=297540

Reviewed by Joshua Hoffman.

Whether or not something is a table can change depending on the role or renderer it has. Today we deal with this by
destroying and recreating the object, which harms performance by forcing us to create new AXIsolatedObjects, update
children of other objects, etc. It's also harmful to ATs, who lose what was otherwise a valid reference to a UIElement,
which can cause various bad behavior such as unexpected focus / AT cursor movement. Finally, we only do this when the
role changes, not when the renderer changes, so the workaround is not comprehensive.

With this commit, we no longer have this concern, as AccessibilityTable is deleted, and functionality moved into
AccessibilityRenderObject / AccessibilityNodeObject.

This commit adds AXObjectRareData to hold onto table-related state. This follows pretty closely to the pattern of
NodeRareData.

* LayoutTests/accessibility/mac/aria-grid-with-strange-hierarchy-expected.txt:
* LayoutTests/accessibility/mac/aria-grid-with-strange-hierarchy.html:
Change expected behavior to include the first row, which is a better experience for ATs and a valid interpretation of
this markup.
* LayoutTests/accessibility/multiple-attribute-change-crash-expected.txt:
* LayoutTests/accessibility/multiple-attribute-change-crash.html:
* Source/WebCore/SaferCPPExpectations/UncheckedCallArgsCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncheckedLocalVarsCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncountedLambdaCapturesCheckerExpectations:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/accessibility/AXCoreObject.cpp:
(WebCore::AXCoreObject::unignoredChildren):
(WebCore::AXCoreObject::firstUnignoredChild):
(WebCore::AXCoreObject::selectedRows):
(WebCore::AXCoreObject::columnHeader):
* Source/WebCore/accessibility/AXCoreObject.h:
(WebCore::Accessibility::exposedTableAncestor):
* Source/WebCore/accessibility/AXLogger.cpp:
(WebCore::AXLogger::log):
(WebCore::operator<<):
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::createObjectFromRenderer):
(WebCore::AXObjectCache::createFromNode):
(WebCore::AXObjectCache::handleChildrenChanged):
(WebCore::AXObjectCache::handleRecomputeCellSlots):
(WebCore::AXObjectCache::handleRowCountChanged):
(WebCore::AXObjectCache::handleRoleChanged):
(WebCore::AXObjectCache::handleAttributeChange):
(WebCore::AXObjectCache::performDeferredCacheUpdate):
(WebCore::AXObjectCache::deferRecomputeTableCellSlots):
(WebCore::isAccessibilityTable): Deleted.
(WebCore::isAccessibilityARIATable): Deleted.
(WebCore::isTableOrRowRole): Deleted.
* Source/WebCore/accessibility/AXObjectCache.h:
* Source/WebCore/accessibility/AXObjectRareData.cpp: Added.
(WebCore::AXObjectRareData::resetChildrenDependentTableFields):
* Source/WebCore/accessibility/AXObjectRareData.h: Added.
(WebCore::AXObjectRareData::tableRows const):
(WebCore::AXObjectRareData::tableColumns const):
(WebCore::AXObjectRareData::rowCount const):
(WebCore::AXObjectRareData::columnCount const):
(WebCore::AXObjectRareData::appendColumn):
(WebCore::AXObjectRareData::appendRow):
(WebCore::AXObjectRareData::isExposableTable const):
(WebCore::AXObjectRareData::setIsExposableTable):
(WebCore::AXObjectRareData::tableHeaderContainer const):
(WebCore::AXObjectRareData::setTableHeaderContainer):
(WebCore::AXObjectRareData::cellSlots const):
(WebCore::AXObjectRareData::mutableCellSlots):
* Source/WebCore/accessibility/AXSearchManager.cpp:
(WebCore::AXSearchManager::matchForSearchKeyAtIndex):
(WebCore::appendChildrenToArray):
* Source/WebCore/accessibility/AXTableHelpers.cpp: Added.
(WebCore::AXTableHelpers::appendCaptionTextIfNecessary):
(WebCore::AXTableHelpers::isTableRole):
(WebCore::AXTableHelpers::tableElementIncludingAncestors):
(WebCore::AXTableHelpers::tableElementIndicatesAccessibleTable):
(WebCore::AXTableHelpers::tableSectionIndicatesAccessibleTable):
(WebCore::AXTableHelpers::styleFrom):
(WebCore::AXTableHelpers::isDataTableWithTraversal):
* Source/WebCore/accessibility/AXTableHelpers.h: Added.
* Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
(WebCore::AccessibilityNodeObject::init):
(WebCore::AccessibilityNodeObject::determineAccessibilityRole):
(WebCore::AccessibilityNodeObject::determineAccessibilityRoleFromNode const):
(WebCore::AccessibilityNodeObject::clearChildren):
(WebCore::AccessibilityNodeObject::addChildren):
(WebCore::AccessibilityNodeObject::isMultiSelectable const):
(WebCore::AccessibilityNodeObject::cellForColumnAndRow):
(WebCore::AccessibilityNodeObject::rareDataWithCleanTableChildren):
(WebCore::AccessibilityNodeObject::cells):
(WebCore::AccessibilityNodeObject::columnCount):
(WebCore::AccessibilityNodeObject::rowCount):
(WebCore::AccessibilityNodeObject::cellSlots):
(WebCore::AccessibilityNodeObject::axRowCount const):
(WebCore::AccessibilityNodeObject::axColumnCount const):
(WebCore::AccessibilityNodeObject::updateRowDescendantRoles):
(WebCore::AccessibilityNodeObject::setCellSlotsDirty):
(WebCore::AccessibilityNodeObject::headerContainer):
(WebCore::AccessibilityNodeObject::columns):
(WebCore::AccessibilityNodeObject::rows):
(WebCore::AccessibilityNodeObject::isDataTable const):
(WebCore::AccessibilityNodeObject::rowHeaders):
(WebCore::AccessibilityNodeObject::visibleRows):
(WebCore::AccessibilityNodeObject::addTableChildrenAndCellSlots):
(WebCore::AccessibilityNodeObject::computeCellSlots):
(WebCore::AccessibilityNodeObject::recomputeIsExposableIfNecessary):
(WebCore::AccessibilityNodeObject::isTable const):
(WebCore::AccessibilityNodeObject::labelText const):
(WebCore::shouldUseAccessibilityObjectInnerText):
* Source/WebCore/accessibility/AccessibilityNodeObject.h:
(WebCore::AccessibilityNodeObject::computeIsTableExposableThroughAccessibility const):
* Source/WebCore/accessibility/AccessibilityObject.cpp:
(WebCore::AccessibilityObject::init):
(WebCore::AccessibilityObject::insertChild):
(WebCore::AccessibilityObject::ensureRareData):
* Source/WebCore/accessibility/AccessibilityObject.h:
(WebCore::AccessibilityObject::hasRareData const):
(WebCore::AccessibilityObject::rareData const):
(WebCore::AccessibilityObject::needsRareData const):
(WebCore::AccessibilityObject::isAccessibilitySVGObjectInstance const):
(WebCore::AccessibilityObject::isAriaTable const):
(WebCore::AccessibilityObject::recomputeIsExposableIfNecessary):
(WebCore::AccessibilityObject::ensureCellIndexesUpToDate):
* Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
(WebCore::AccessibilityRenderObject::computeIsIgnored const):
(WebCore::AccessibilityRenderObject::determineAccessibilityRole):
(WebCore::AccessibilityRenderObject::addChildren):
* Source/WebCore/accessibility/AccessibilityTable.cpp: Removed.
* Source/WebCore/accessibility/AccessibilityTable.h: Removed.
* Source/WebCore/accessibility/AccessibilityTableCell.cpp:
(WebCore::AccessibilityTableCell::parentTable const):
(WebCore::AccessibilityTableCell::isExposedTableCell const):
(WebCore::AccessibilityTableCell::determineAccessibilityRole):
* Source/WebCore/accessibility/AccessibilityTableCell.h:
* Source/WebCore/accessibility/AccessibilityTableColumn.cpp:
(WebCore::AccessibilityTableColumn::addChildren):
* Source/WebCore/accessibility/AccessibilityTableHeaderContainer.cpp:
(WebCore::AccessibilityTableHeaderContainer::addChildren):
* Source/WebCore/accessibility/AccessibilityTableHeaderContainer.h:
* Source/WebCore/accessibility/AccessibilityTableRow.cpp:
(WebCore::AccessibilityTableRow::isTableRow const):
(WebCore::AccessibilityTableRow::parentTable const):
(WebCore::AccessibilityTableRow::disclosedRows):
(WebCore::AccessibilityTableRow::disclosedByRow const):
* Source/WebCore/accessibility/AccessibilityTableRow.h:
* Source/WebCore/accessibility/ios/WebAccessibilityObjectWrapperIOS.mm:
(-[WebAccessibilityObjectWrapper tableParent]):
(-[WebAccessibilityObjectWrapper accessibilityHeaderElements]):
(-[WebAccessibilityObjectWrapper accessibilityElementForRow:andColumn:]):
(-[WebAccessibilityObjectWrapper accessibilityRowCount]):
(-[WebAccessibilityObjectWrapper accessibilityColumnCount]):
(-[WebAccessibilityObjectWrapper accessibilityARIARowCount]):
(-[WebAccessibilityObjectWrapper accessibilityARIAColumnCount]):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.cpp:
(WebCore::AXIsolatedObject::isAccessibilityTableInstance const): Deleted.
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.h:
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:
(WebCore::AXIsolatedTree::updateDependentProperties):
(WebCore::createIsolatedObjectData):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.h:
* Source/WebCore/accessibility/mac/AXObjectCacheMac.mm:
(WebCore::AXObjectCache::postPlatformNotification):
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperBase.mm:
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.mm:
(-[WebAccessibilityObjectWrapper _additionalAccessibilityAttributeNames:]):
(-[WebAccessibilityObjectWrapper accessibilityAttributeNames]):
(-[WebAccessibilityObjectWrapper accessibilityAttributeValue:]):
(-[WebAccessibilityObjectWrapper accessibilityParameterizedAttributeNames]):
(-[WebAccessibilityObjectWrapper _accessibilitySetValue:forAttribute:]):
(-[WebAccessibilityObjectWrapper accessibilityAttributeValue:forParameter:]):

Canonical link: https://commits.webkit.org/298972@main

iangrunert commented on Source/WebCore/accessibility/AXTableHelpers.h:
```diff
@@ -0,0 +1,54 @@
+
+// This value is what will be used if AccessibilityTableCell determines the cell
+// should not be treated as a cell (e.g. because it is in a layout table).
+static constexpr AccessibilityRole layoutTableCellRole = AccessibilityRole::TextGroup;
```
AccessibilityRole::TextGroup isn't available here in non-unified builds; AccessibilityRole is forward declared on line 30.

twilco replied:
Ahhh, yes, sorry 😞

twilco replied:
If you haven't already fixed this one — one probably easy option would be to move this to AXCoreObject.h, which is where `AccessibilityRole` is defined. Ideally we would separate `AccessibilityRole` into its own header that we could include here, but I'll do that later.
